File: signal_engine.py
from __future__ import annotations

from dataclasses import dataclass, field

import pandas as pd

import config
import indicators as ind
import patterns as pat

BULLISH_PATTERNS = {"bullish_engulfing", "morning_star", "hammer"}
BEARISH_PATTERNS = {"bearish_engulfing", "evening_star"}


@dataclass
class SignalResult:
    symbol: str
    signal: str
    close: float
    atr: float
    adx: float
    pattern: str | None
    confirmations_bull: list[str] = field(default_factory=list)
    confirmations_bear: list[str] = field(default_factory=list)
    stop_loss: float | None = None
    take_profit: float | None = None
    position_size: float | None = None
    rationale: str = ""


def compute_indicators(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    df["ema_fast"] = ind.ema(df["close"], config.EMA_FAST)
    df["ema_slow"] = ind.ema(df["close"], config.EMA_SLOW)
    df["ema_trend"] = ind.ema(df["close"], config.EMA_TREND)
    df["rsi"] = ind.rsi(df["close"], config.RSI_PERIOD)
    macd_line, signal_line, hist = ind.macd(
        df["close"], config.MACD_FAST, config.MACD_SLOW, config.MACD_SIGNAL
    )
    df["macd"], df["macd_signal"], df["macd_hist"] = macd_line, signal_line, hist
    df["adx"] = ind.adx(df, config.ADX_PERIOD)
    df["atr"] = ind.atr(df, config.ATR_PERIOD)
    df["volume_sma"] = df["volume"].rolling(20).mean()
    return df


def _confluence_at(df: pd.DataFrame, i: int) -> tuple[list[str], list[str], str | None]:
    last, prev = df.iloc[i], df.iloc[i - 1]
    pattern = pat.detect_pattern_at(df, i)

    bull: list[str] = []
    bear: list[str] = []

    if last["ema_fast"] > last["ema_slow"] and prev["ema_fast"] <= prev["ema_slow"]:
        bull.append("croisement_ema_haussier")
    if last["ema_fast"] < last["ema_slow"] and prev["ema_fast"] >= prev["ema_slow"]:
        bear.append("croisement_ema_baissier")
    if last["ema_fast"] > last["ema_slow"] > last["ema_trend"]:
        bull.append("structure_ema_haussiere")
    if last["ema_fast"] < last["ema_slow"] < last["ema_trend"]:
        bear.append("structure_ema_baissiere")

    if pattern in BULLISH_PATTERNS:
        bull.append(f"pattern_{pattern}")
    if pattern in BEARISH_PATTERNS:
        bear.append(f"pattern_{pattern}")

    if 40 < last["rsi"] < 70 and last["rsi"] > prev["rsi"] and prev["rsi"] < 45:
        bull.append("rsi_sortie_survente")
    if 30 < last["rsi"] < 60 and last["rsi"] < prev["rsi"] and prev["rsi"] > 55:
        bear.append("rsi_sortie_surachat")

    if last["macd_hist"] > 0 and prev["macd_hist"] <= 0:
        bull.append("macd_croisement_haussier")
    if last["macd_hist"] < 0 and prev["macd_hist"] >= 0:
        bear.append("macd_croisement_baissier")

    if last["volume"] > last["volume_sma"] * 1.2:
        (bull if last["close"] > last["open"] else bear).append("volume_confirme")

    return bull, bear, pattern


def signal_from_row(symbol: str, df: pd.DataFrame, i: int) -> SignalResult:
    last = df.iloc[i]
    bull, bear, pattern = _confluence_at(df, i)
    trending = last["adx"] > config.ADX_TREND_THRESHOLD

    result = SignalResult(
        symbol=symbol,
        signal="",
        close=last["close"],
        atr=last["atr"],
        adx=last["adx"],
        pattern=pattern,
        confirmations_bull=bull,
        confirmations_bear=bear,
    )

    if trending and len(bull) >= config.MIN_CONFIRMATIONS and len(bull) > len(bear):
        result.signal = "ACHAT"
    elif trending and len(bear) >= config.MIN_CONFIRMATIONS and len(bear) > len(bull):
        result.signal = "VENTE"
    elif not trending and pattern is None:
        result.signal = "CALME"
    else:
        result.signal = "A_SURVEILLER"

    result.rationale = _build_rationale(result)
    return result


def evaluate(symbol: str, df: pd.DataFrame) -> SignalResult:
    df = compute_indicators(df)
    return signal_from_row(symbol, df, len(df) - 1)


def _build_rationale(r: SignalResult) -> str:
    if r.signal == "ACHAT":
        return f"Confluence haussiere ({', '.join(r.confirmations_bull)}), ADX={r.adx:.1f}"
    if r.signal == "VENTE":
        return f"Confluence baissiere ({', '.join(r.confirmations_bear)}), ADX={r.adx:.1f}"
    if r.signal == "CALME":
        return f"ADX={r.adx:.1f} sous le seuil de tendance, aucune figure significative"
    return (
        f"Signaux insuffisants ou contradictoires "
        f"(haussier={len(r.confirmations_bull)}, baissier={len(r.confirmations_bear)}), a confirmer"
    )

File: pwa/js/signalEngine.js
// Port de signal_engine.py + risk_manager.py

function computeIndicators(candles) {
  const closes = candles.map((c) => c.close);
  const volumes = candles.map((c) => c.volume);
  const emaFastArr = ema(closes, CONFIG.emaFast);
  const emaSlowArr = ema(closes, CONFIG.emaSlow);
  const emaTrendArr = ema(closes, CONFIG.emaTrend);
  const rsiArr = rsi(closes, CONFIG.rsiPeriod);
  const { macdLine, signalLine, histogram } = macd(closes, CONFIG.macdFast, CONFIG.macdSlow, CONFIG.macdSignal);
  const adxArr = adx(candles, CONFIG.adxPeriod);
  const atrArr = atr(candles, CONFIG.atrPeriod);
  const volumeSmaArr = sma(volumes, 20);

  return candles.map((c, i) => ({
    ...c,
    emaFast: emaFastArr[i],
    emaSlow: emaSlowArr[i],
    emaTrend: emaTrendArr[i],
    rsi: rsiArr[i],
    macdHist: histogram[i],
    adx: adxArr[i],
    atr: atrArr[i],
    volumeSma: volumeSmaArr[i],
  }));
}

function confluenceAt(data, i) {
  const last = data[i];
  const prev = data[i - 1];
  const pattern = detectPatternAt(data, i);
  const bull = [];
  const bear = [];

  if (last.emaFast > last.emaSlow && prev.emaFast <= prev.emaSlow) bull.push("croisement_ema_haussier");
  if (last.emaFast < last.emaSlow && prev.emaFast >= prev.emaSlow) bear.push("croisement_ema_baissier");
  if (last.emaFast > last.emaSlow && last.emaSlow > last.emaTrend) bull.push("structure_ema_haussiere");
  if (last.emaFast < last.emaSlow && last.emaSlow < last.emaTrend) bear.push("structure_ema_baissiere");

  if (BULLISH_PATTERNS.has(pattern)) bull.push(`pattern_${pattern}`);
  if (BEARISH_PATTERNS.has(pattern)) bear.push(`pattern_${pattern}`);

  if (last.rsi > 40 && last.rsi < 70 && last.rsi > prev.rsi && prev.rsi < 45) bull.push("rsi_sortie_survente");
  if (last.rsi > 30 && last.rsi < 60 && last.rsi < prev.rsi && prev.rsi > 55) bear.push("rsi_sortie_surachat");

  if (last.macdHist > 0 && prev.macdHist <= 0) bull.push("macd_croisement_haussier");
  if (last.macdHist < 0 && prev.macdHist >= 0) bear.push("macd_croisement_baissier");

  if (last.volume > last.volumeSma * 1.2) {
    (last.close > last.open ? bull : bear).push("volume_confirme");
  }

  return { bull, bear, pattern };
}

function buildRationale(signal, bull, bear, adxVal) {
  if (signal === "ACHAT") return `Confluence haussiere (${bull.join(", ")}), ADX=${adxVal.toFixed(1)}`;
  if (signal === "VENTE") return `Confluence baissiere (${bear.join(", ")}), ADX=${adxVal.toFixed(1)}`;
  if (signal === "CALME") return `ADX=${adxVal.toFixed(1)} sous le seuil de tendance, aucune figure significative`;
  return `Signaux insuffisants ou contradictoires (haussier=${bull.length}, baissier=${bear.length}), a confirmer`;
}

function signalFromRow(symbol, data, i) {
  const last = data[i];
  const { bull, bear, pattern } = confluenceAt(data, i);
  const trending = last.adx > CONFIG.adxTrendThreshold;

  let signal;
  if (trending && bull.length >= CONFIG.minConfirmations && bull.length > bear.length) signal = "ACHAT";
  else if (trending && bear.length >= CONFIG.minConfirmations && bear.length > bull.length) signal = "VENTE";
  else if (!trending && pattern === null) signal = "CALME";
  else signal = "A_SURVEILLER";

  return {
    symbol,
    signal,
    close: last.close,
    atr: last.atr,
    adx: last.adx,
    pattern,
    confirmationsBull: bull,
    confirmationsBear: bear,
    rationale: buildRationale(signal, bull, bear, last.adx),
    stopLoss: null,
    takeProfit: null,
    positionSize: null,
  };
}

function evaluate(symbol, candles) {
  const data = computeIndicators(candles);
  return signalFromRow(symbol, data, data.length - 1);
}

function applyRiskManagement(result, capital = CONFIG.capitalTotal) {
  if (result.signal !== "ACHAT" && result.signal !== "VENTE") return result;
  const stopDistance = CONFIG.atrStopMultiplier * result.atr;
  if (stopDistance <= 0) return result;

  if (result.signal === "ACHAT") {
    result.stopLoss = result.close - stopDistance;
    result.takeProfit = result.close + stopDistance * CONFIG.takeProfitRR;
  } else {
    result.stopLoss = result.close + stopDistance;
    result.takeProfit = result.close - stopDistance * CONFIG.takeProfitRR;
  }

  const riskAmount = capital * CONFIG.riskPerTradePct;
  result.positionSize = riskAmount / stopDistance;
  const maxNotional = capital * CONFIG.maxExposurePerAssetPct;
  if (result.positionSize * result.close > maxNotional) {
    result.positionSize = maxNotional / result.close;
  }
  return result;
}
